### PR TITLE
Add support for Python 3

### PR DESCRIPTION
**Pull request by thet**
This is primarily the output of ``python-modernize -x import -wn`` and untested....

### CODE REVIEW

**Comment by reinhardt** [comment]
Tests are now green for me locally. We may want to set up travis or jenkins.

File: src/collective/vdexvocabulary/term.py [hunk 1]
```diff
@@ -28,7 +28,7 @@ def __init__(self, value, token=None, title=None, description=None, related=[]):
     def __repr__(self):
         rep = "<%s '%s' at %s>" % (
             self.__class__.__name__,
-            self.token,
+            self.token.decode("utf-8"),
```
Comment (ale-rt) [inline src/collective/vdexvocabulary/term.py]: This should be safe_nativestring ot it might break.

Comment (reinhardt) [reply src/collective/vdexvocabulary/term.py]: I actually "fixed" the wrong spot. `__init__` must make sure that the token is always a nativestring (as specified by ITokenizedTerm). Then no further fiddling is needed in `__repr__`.

File: src/collective/vdexvocabulary/term.py [hunk 2]
```diff
@@ -1,21 +1,21 @@
...
 
-    implements(ITokenizedTerm)
-
     def __init__(self, value, token=None, title=None, description=None, related=[]):
         self.value = value
         if token is None:
             token = value
-        if isinstance(token, unicode):
-            self.token = token.encode("utf-8")
+        if six.PY3 and isinstance(token, bytes):
```
Comment (ale-rt) [inline src/collective/vdexvocabulary/term.py]: I would use `safe_nativestring` here or some equivalent thing.

File: src/collective/vdexvocabulary/term.rst [hunk 3]
```diff
@@ -8,19 +8,30 @@ We use some german umlauts here, but repr is needed in a doctest :/
 
 ::
 
-    >>> term = VdexTerm(u'f\xfc\xfc')
+    >>> term = VdexTerm('füü')
     >>> term
     <VdexTerm 'füü' at 0x...>
 
-    >>> isinstance(term.value, unicode)
+    >>> isinstance(term.value, str)
```
Comment (ale-rt) [inline src/collective/vdexvocabulary/term.rst]: Value should be `six.text_type`

File: src/collective/vdexvocabulary/term.rst [hunk 4]
```diff
@@ -8,19 +8,30 @@ We use some german umlauts here, but repr is needed in a doctest :/
 
 ::
 
-    >>> term = VdexTerm(u'f\xfc\xfc')
+    >>> term = VdexTerm('füü')
```
Comment (ale-rt) [inline src/collective/vdexvocabulary/term.rst]: Should be:
```
term = VdexTerm(u'füü')
```

File: src/collective/vdexvocabulary/term.rst [hunk 5]
```diff
@@ -8,19 +8,30 @@ We use some german umlauts here, but repr is needed in a doctest :/
...
     True
 
     >>> term.title is None
     True
 
+Passing bytes as a token should work, too.
+
+::
+
+    >>> term = VdexTerm('füü', token='füü'.encode("utf-8"))
```
Comment (ale-rt) [inline src/collective/vdexvocabulary/term.rst]: Should be:
```
>>> term = VdexTerm('füü', token=b'füü')
```

File: src/collective/vdexvocabulary/tests.py [hunk 6]
```diff
@@ -25,7 +22,6 @@ def test_suite():
                 filename,
                 optionflags=optionflags,
                 globs={
-                    "interact": interact,
```
Comment (ale-rt) [inline src/collective/vdexvocabulary/tests.py:28]: I do not know why this was removed :)

File: src/collective/vdexvocabulary/treevocabulary.py [hunk 7]
```diff
@@ -21,15 +23,23 @@ def domain(self):
...
+        number=None,
     ):
         # msgid is always kind|termidentifer
         # i.e.: caption|prod.1 or description|prod.45
         kind, msgid = msgid.split("|")
 
         # handle default
         if default is None:
-            default = unicode(msgid)
+            default = six.text_type(msgid)
```
Comment (ale-rt) [inline src/collective/vdexvocabulary/treevocabulary.py]: I would use `safe_unicode`

File: src/collective/vdexvocabulary/treevocabulary.py [hunk 8]
```diff
@@ -51,13 +61,13 @@ def translate(
 
         # find out what the target language should be
         if target_language is None and context is not None:
-            langs = translations.keys()
+            langs = list(translations.keys())
```
Comment (ale-rt) [inline src/collective/vdexvocabulary/treevocabulary.py]: `list(translations)` should be enough

File: src/collective/vdexvocabulary/treevocabulary.py [hunk 9]
```diff
@@ -51,13 +61,13 @@ def translate(
...
             negotiator = getUtility(INegotiator)
             target_language = negotiator.getLanguage(langs, context)
 
         # fetch matching translation or default
         message = translations.get(target_language, default)
-        if not isinstance(message, unicode):
+        if not isinstance(message, six.text_type):
```
Comment (ale-rt) [inline src/collective/vdexvocabulary/treevocabulary.py]: I would use `safe_nativesting`.

File: src/collective/vdexvocabulary/treevocabulary.py [hunk 10]
```diff
@@ -106,12 +116,12 @@ def _createTermTree(self, element, terms, _):
                 # i18n message id for title/caption
                 _(
                     "caption|%s" % identifier,
-                    default=default_title or identifier.decode("utf-8"),
+                    default=default_title or identifier,
```
Comment (ale-rt) [inline src/collective/vdexvocabulary/treevocabulary.py]: This should be `safe_unicode(default_title or identifier)` it seems (before it was decoded).

File: src/collective/vdexvocabulary/treevocabulary.py [hunk 11]
```diff
@@ -106,12 +116,12 @@ def _createTermTree(self, element, terms, _):
...
                 ),
                 # i18n message id for description
                 _(
                     "description|%s" % identifier,
-                    default=default_description or identifier.decode("utf-8"),
+                    default=default_description or identifier,
```
Comment (ale-rt) [inline src/collective/vdexvocabulary/treevocabulary.py]: This should be `safe_unicode(default_title or identifier)` it seems (before it was decoded).

File: src/collective/vdexvocabulary/treevocabulary.rst [hunk 12]
```diff
@@ -54,13 +54,13 @@ check translations::
 
     >>> from zope.i18n import translate
     >>> translate(vdex_vocab.term_by_value['nwe'].title)
-    u'North-west of Europe'
+    'North-west of Europe'
```
Comment (ale-rt) [inline src/collective/vdexvocabulary/treevocabulary.rst:57]: This should return unicode on Py2.7... I think something like https://github.com/plone/plone.indexer/blob/50a0ead2afba37241e21db1ed04985f06ec0f22c/plone/indexer/tests.py#L25 might be needed

File: src/collective/vdexvocabulary/vocabulary.py [hunk 13]
```diff
@@ -160,7 +160,7 @@ def getRelations(self, context, lang, items):
                     + ")"
                 )
 
-            if sourceTerm not in out.keys():
+            if sourceTerm not in list(out.keys()):
```
Comment (ale-rt) [inline src/collective/vdexvocabulary/vocabulary.py]: `list(out)` should be enough

**Review by ale-rt — CHANGES_REQUESTED** [review]

File: src/collective/vdexvocabulary/tests.py (hunk 6 above)
Comment (thet) [reply src/collective/vdexvocabulary/tests.py:28]: interact is only to have a python shell in doctests. mostly useful for developing or debugging doctests. should not a necessary dependency. https://pypi.org/project/interlude/

**Review by ale-rt — APPROVED** [review]

File: src/collective/vdexvocabulary/term.py [hunk 14]
```diff
@@ -1,24 +1,33 @@
...
 from zope.schema.vocabulary import getVocabularyRegistry
+import six
 
+try:
+    from Products.CMFPlone.utils import safe_nativestring
+except ImportError:
+    if six.PY2:
+        from Products.CMFPlone.utils import safe_encode as safe_nativestring
+    else:
+        from Products.CMFPlone.utils import safe_unicode as safe_nativestring
```
Comment (reinhardt) [inline src/collective/vdexvocabulary/term.py]: We don't have a dependency on Products.CMFPlone. I think it's intentional that you can install the package in a pure Zope environment. If we want to drop that option, we at least need to declare the dependency on Plone in setup.py.

File: src/collective/vdexvocabulary/treevocabulary.py [hunk 15]
```diff
@@ -9,6 +9,13 @@
 
 import imsvdex.vdex
 import os
+import six
+
+try:
+    from Products.CMFPlone.utils import safe_text
+except ImportError:
+    # Plone 5.1
+    from Products.CMFPlone.utils import safe_unicode as safe_text
```
Comment (reinhardt) [inline src/collective/vdexvocabulary/treevocabulary.py]: See above about the dependency on Plone

**Review by reinhardt — CHANGES_REQUESTED** [review]
In general I'm not a fan of just using the safe_* helper methods. Sometimes if you get a wrong input type it can point to a bug somewhere else, which you would only cover up by quietly converting and trying to use it anyway. That said, we don't have the safe_* methods available here because we don't depend on CMFPlone. Either we need to redefine them or declare the dependency. I think the original intention was to allow installing in a pure Zope instance, but it may well be that no one is doing that anyway...

**Comment by ale-rt** [comment]
Good points! I will recheck :) Thanks for your review!

File: src/collective/vdexvocabulary/term.py (hunk 14 above)
Comment (ale-rt) [reply src/collective/vdexvocabulary/term.py]: I replaced `safe_nativestring` with `six.ensure_str`.

File: src/collective/vdexvocabulary/treevocabulary.py (hunk 15 above)
Comment (ale-rt) [reply src/collective/vdexvocabulary/treevocabulary.py]: I replaced `safe_text` with `six.ensure_text`.

**Review by reinhardt — APPROVED** [review]
I didn't know six had these methods! Cool!

**Comment by ale-rt** [comment]
> I didn't know six had these methods! Cool!

Me too, they are a quite recent addition anyway :)

**Pull request merged**